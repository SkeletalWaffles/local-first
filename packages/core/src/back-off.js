// @flow

const backOff = (
    fn: () => Promise<boolean>,
    wait: number = 200,
    rate: number = 1.5,
    initialWait: number = wait,
) => {
    fn()
        .catch(err => false)
        .then(succeeded => {
            if (succeeded) {
                return;
            } else if (globalThis.document != undefined) {
                const tid = setTimeout(() => {
                    document.removeEventListener('visibilitychange', listener, false);
                    backOff(fn, wait * rate, rate, initialWait);
                }, wait);

                const listener = () => {
                    if (!document.hidden) {
                        document.removeEventListener('visibilitychange', listener, false);
                        clearTimeout(tid);
                        backOff(fn, initialWait, rate, initialWait);
                    }
                };

                if (wait > 1000) {
                    document.addEventListener('visibilitychange', listener, false);
                }
            } else {
                backOff(fn, wait * rate, rate, initialWait);
            }
        });
};

// function handleVisibilityChange() {
//   if (document.hidden) {
//     pauseSimulation();
//   } else  {
//     startSimulation();
//   }
// }

// document.addEventListener("visibilitychange", handleVisibilityChange, false);

export default backOff;
