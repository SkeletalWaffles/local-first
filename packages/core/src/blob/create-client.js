// @flow

import type {
    Client,
    Collection,
    PeerChange,
    Blob,
    PersistentClock,
    FullPersistence,
    BlobNetworkCreator,
} from '../types';
import { peerTabAwareNetwork } from '../peer-tabs';
import { type Schema } from '../../../packages/nested-object-crdt/src/schema.js';
import * as hlc from '../../../packages/hybrid-logical-clock';
import deepEqual from 'fast-deep-equal';

import {
    newCollection,
    getCollection,
    onCrossTabChanges,
    type CRDTImpl,
    type CollectionState,
} from '../shared';

const genId = () =>
    Math.random()
        .toString(36)
        .slice(2);

import { type ClientMessage, type ServerMessage } from '../server';

export const fullMaxStamp = function<Delta, Data>(
    crdt: CRDTImpl<Delta, Data>,
    full: Blob<Data>,
) {
    let maxStamp = null;
    Object.keys(full).forEach(colid => {
        Object.keys(full[colid]).forEach(key => {
            const latest = crdt.maxStamp(full[colid][key]);
            if (latest && (!maxStamp || latest > maxStamp)) {
                maxStamp = latest;
            }
        });
    });
    return maxStamp;
};

export const updateCacheAndNotify = function<Delta, Data>(
    state: { [key: string]: CollectionState<Data, any> },
    crdt: CRDTImpl<Delta, Data>,
    changedIds: { [key: string]: Array<string> },
    blob: Blob<Data>,
    sendCrossTabChanges: PeerChange => mixed,
) {
    Object.keys(changedIds).forEach(colid => {
        const col = state[colid];
        const data = blob[colid];
        if (col.listeners.length) {
            const changes = changedIds[colid].map(id => ({
                id,
                value: crdt.value(data[id]),
            }));
            changedIds[colid].forEach(id => {
                state[colid].cache[id] = data[id];
            });
            col.listeners.forEach(listener => {
                listener(changes);
            });
        }
        changedIds[colid].forEach(id => {
            // Only update the cache if the node has already been cached
            // Umm is this necessary though?
            if (state[colid].cache[id] || col.itemListeners[id]) {
                state[colid].cache[id] = data[id];
            }
            if (col.itemListeners[id]) {
                col.itemListeners[id].forEach(fn => fn(crdt.value(data[id])));
            }
        });

        if (changedIds[colid].length) {
            console.log(
                'Broadcasting to client-level listeners',
                changedIds[colid],
            );
            sendCrossTabChanges({
                col: colid,
                nodes: changedIds[colid],
            });
        }
    });
};

function createClient<Delta, Data, SyncStatus>(
    crdt: CRDTImpl<Delta, Data>,
    schemas: { [colid: string]: Schema },
    clock: PersistentClock,
    persistence: FullPersistence,
    createNetwork: BlobNetworkCreator<Data, SyncStatus>,
): Client<SyncStatus> {
    const state: { [key: string]: CollectionState<Data, any> } = {};
    persistence.collections.forEach(id => (state[id] = newCollection()));

    const network = peerTabAwareNetwork(
        (msg: PeerChange) => {
            return onCrossTabChanges(
                crdt,
                persistence,
                state[msg.col],
                msg.col,
                msg.nodes,
            );
        },
        createNetwork(
            persistence.getFull,
            async (full, etag, sendCrossTabChanges) => {
                const max = fullMaxStamp(crdt, full);
                if (max) {
                    clock.recv(max);
                }
                const result = await persistence.mergeFull<Data>(
                    full,
                    etag,
                    crdt.merge,
                );
                if (!result) {
                    return null;
                }
                const { merged, changedIds } = result;
                updateCacheAndNotify(
                    state,
                    crdt,
                    changedIds,
                    merged.blob,
                    sendCrossTabChanges,
                );
                return merged;
            },
            persistence.updateMeta,
        ),
    );

    return {
        sessionId: clock.now.node,
        setDirty: network.setDirty,
        getStamp: clock.get,
        getCollection<T>(colid: string) {
            return getCollection(
                colid,
                crdt,
                persistence,
                state[colid],
                clock.get,
                network.setDirty,
                network.sendCrossTabChanges,
                schemas[colid],
            );
        },
        onSyncStatus(fn) {
            network.onSyncStatus(fn);
        },
        getSyncStatus() {
            return network.getSyncStatus();
        },
    };
}

export default createClient;
