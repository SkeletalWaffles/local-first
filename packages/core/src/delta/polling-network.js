// @flow

import type {
    Persistence,
    Network,
    ClockPersist,
    DeltaPersistence,
    FullPersistence,
    NetworkCreator,
} from '../types';
import type { HLC } from '@local-first/hybrid-logical-clock';
import * as hlc from '@local-first/hybrid-logical-clock';
import deepEqual from 'fast-deep-equal';
import { type ClientMessage, type ServerMessage } from '../server';

type SyncStatus = { status: 'connected' } | { status: 'disconnected' };

import { peerTabAwareSync } from '../peer-tabs';
import poller from '../poller';
import backOff from '../back-off';
import { debounce } from '../debounce';

// Ok the part where we get very specific
const syncFetch = async function<Delta, Data>(
    url: string,
    sessionId: string,
    getMessages: (
        reconnected: boolean,
    ) => Promise<Array<ClientMessage<Delta, Data>>>,
    onMessages: (Array<ServerMessage<Delta, Data>>) => Promise<mixed>,
) {
    const messages = await getMessages(true);
    console.log('sync:messages', messages);
    // console.log('messages', messages);
    const res = await fetch(`${url}?sessionId=${sessionId}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(messages),
    });
    if (res.status !== 200) {
        throw new Error(`Unexpected status ${res.status}`);
    }
    const data = await res.json();
    console.log('sync:data', data);
    await onMessages(data);
};

const createPollingNetwork = <Delta, Data>(
    url: string,
): NetworkCreator<Delta, Data, SyncStatus> => (
    sessionId,
    getMessages,
    handleMessages,
): Network<SyncStatus> => {
    return {
        initial: { status: 'disconnected' },
        createSync: (sendCrossTabChange, updateStatus) => {
            console.log('Im the leader (polling)');
            const poll = poller(
                3 * 1000,
                () =>
                    new Promise(res => {
                        backOff(() =>
                            syncFetch(url, sessionId, getMessages, messages =>
                                handleMessages(messages, sendCrossTabChange),
                            ).then(
                                () => {
                                    updateStatus({ status: 'connected' });
                                    res();
                                    return true;
                                },
                                err => {
                                    console.error('Failed to sync polling');
                                    console.error(err.stack);
                                    updateStatus({
                                        status: 'disconnected',
                                    });
                                    return false;
                                },
                            ),
                        );
                    }),
            );
            poll();
            return debounce(poll);
        },
    };
};

export default createPollingNetwork;
