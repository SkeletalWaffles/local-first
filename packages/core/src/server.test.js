// @flow

import type { ClientMessage, ServerMessage } from './server';

import * as ncrdt from '../../nested-object-crdt/src/new';
import * as rich from '../../rich-text-crdt/index';
import makePersistence from '../../idb/src/delta-mem';
import * as client from './delta/create-client';
import * as server from './server';
import setupServerPersistence from './memory-persistence';
import { PersistentClock, inMemoryClockPersist } from './persistent-clock';
import { type CRDTImpl, getCollection } from './shared';
import type { ServerState } from './server';

const otherMerge = (v1, m1, v2, m2) => {
    return { value: rich.merge(v1, v2), meta: null };
};
const applyOtherDelta = (text: rich.CRDT, meta: null, delta: rich.Delta) => {
    return {
        value: rich.apply(text, delta),
        meta,
    };
};

const newCrdt: CRDTImpl<Delta, Data> = {
    merge: (one, two) => {
        if (!one) return two;
        // $FlowFixMe
        return ncrdt.mergeTwo(one, two, () => {
            throw new Error('nope');
        });
    },
    latestStamp: data => ncrdt.latestStamp(data, () => null),
    value: d => d.value,
    applyDelta: (base, delta) =>
        ncrdt.applyDelta(
            base,
            delta,
            // $FlowFixMe
            applyOtherDelta,
            otherMerge,
        ),
    deltas: {
        ...ncrdt.deltas,
        stamp: data => ncrdt.deltas.stamp(data, () => null),
        apply: (base, delta) =>
            ncrdt.applyDelta(
                base,
                delta,
                // $FlowFixMe
                applyOtherDelta,
                otherMerge,
            ),
    },
    createValue: (value, stamp, getStamp, schema) => {
        return ncrdt.createWithSchema(
            value,
            stamp,
            getStamp,
            schema,
            value => null,
        );
    },
};

/*

server-boot-up
client -> hello
server -> datas
client -> ack
...
client -> new changes
server -> ack, new cursor?
client -> ack
...
client2 -> new changes
server -> ack, new cursor
server -> client2 datas + cursor
client -> ack
client2 -> ack

*/

type Data = ncrdt.CRDT<any, any>;
type Delta = ncrdt.Delta<any, any, rich.Delta>;

const schemas = {
    items: {
        type: 'object',
        attributes: {
            one: 'int',
            three: 'string',
            five: { type: 'object', attributes: { six: 'int' } },
        },
    },
    people: { type: 'object', attributes: { name: 'string', age: 'int' } },
};

const collections = ['items'];
const initialValue = ncrdt.createDeep(
    { one: 2, three: '4', five: { six: 7 } },
    '1',
);
const someMessages = [
    {
        node: 'yes',
        delta: ncrdt.deltas.replace(initialValue),
    },
    {
        node: 'yes',
        delta: ncrdt.deltas.set(
            initialValue,
            ['five', 'six'],
            ncrdt.createDeep(8, '2'),
        ),
    },
];
const expectedValue = { yes: { one: 2, three: '4', five: { six: 8 } } };

const createClient = (sessionId, collections, messages) => {
    const persistence = makePersistence('yolo', collections);
    const state = client.initialState(persistence.collections);
    const clock = new PersistentClock(inMemoryClockPersist());
    // client
    return {
        sessionId,
        collections,
        getMessages(): Promise<Array<ClientMessage<Delta, Data>>> {
            return client.getMessages(persistence, false);
        },
        receive(
            messages: Array<ServerMessage<Delta, Data>>,
        ): Promise<Array<ClientMessage<Delta, Data>>> {
            return client.handleMessages(
                newCrdt,
                persistence,
                messages,
                state,
                clock.recv,
                () => {},
            );
        },
        async getState() {
            const data = {};
            for (let col of collections) {
                data[col] = await this.getCollection(col).loadAll();
            }
            return data;
        },
        getCollection(colid: string) {
            return getCollection(
                colid,
                newCrdt,
                persistence,
                state[colid],
                clock.get,
                () => {},
                () => {},
                schemas[colid],
            );
        },
    };
};

const createServer = deltas => {
    const state: ServerState<Delta, Data> = server.default<Delta, Data>(
        newCrdt,
        setupServerPersistence<Delta, Data>(),
    );
    if (deltas) {
        Object.keys(deltas).forEach(collection => {
            server.onMessage(state, '-none-', {
                type: 'sync',
                collection,
                serverCursor: -1,
                deltas: deltas[collection],
            });
        });
    }

    // server
    return {
        getMessages(sessionId: string): Array<ServerMessage<Delta, Data>> {
            return server.getMessages(state, sessionId);
        },
        receive(
            sessionId: string,
            messages: Array<ClientMessage<Delta, Data>>,
        ): Array<ServerMessage<Delta, Data>> {
            const responses = messages
                .map(message => server.onMessage(state, sessionId, message))
                .filter(Boolean);
            return responses;
        },
        getState(collections: Array<string>) {
            const data = {};
            collections.forEach(cid => {
                data[cid] = {};
                const fulls = {};
                const result = state.persistence.deltasSince(cid, -1, '');
                if (!result) return;
                result.deltas.forEach(({ node, delta }) => {
                    fulls[node] = state.crdt.applyDelta(fulls[node], delta);
                });
                Object.keys(fulls).forEach(key => {
                    data[cid][key] = state.crdt.value(fulls[key]);
                });
            });
            return data;
        },
    };
};

describe('client-server interaction', () => {
    it('Clean both - initial handshake', async () => {
        const client = createClient('a', ['items']);
        const server = createServer();
        const initialClient = await client.getMessages();
        expect(initialClient).toEqual(
            client.collections.map(name => ({
                type: 'sync',
                collection: name,
                serverCursor: undefined,
                deltas: [],
            })),
        );
        expect(server.receive(client.sessionId, initialClient)).toEqual([]);
        const initialServer = server.getMessages(client.sessionId);
        expect(initialServer).toEqual(
            client.collections.map(collection => ({
                type: 'sync',
                collection,
                serverCursor: -1,
                deltas: [],
            })),
        );
        const acks = await client.receive(initialServer);
        expect(acks).toEqual(
            client.collections.map(collection => ({
                type: 'ack',
                collection,
                serverCursor: -1,
            })),
        );
        expect(await client.getMessages()).toEqual([]);
        expect(server.receive(client.sessionId, acks)).toEqual([]);
        expect(server.getMessages(client.sessionId)).toEqual([]);
        expect(await client.getState()).toEqual(
            server.getState(client.collections),
        );
    });

    it('Clean client, server with data should back and forth', async () => {
        const client = createClient('a', ['items']);
        const server = createServer({ items: someMessages });
        // hello world, back and forth
        const acks = server.receive(
            client.sessionId,
            await client.getMessages(),
        );
        const allServer = acks.concat(server.getMessages(client.sessionId));
        console.log(allServer);
        const clientAcks = await client.receive(allServer);
        // expect(clientAcks).toEqual([]);
        // client is ack'ing the server's data, but doesn't have its own messages.
        server.receive(client.sessionId, clientAcks);
        expect(await client.getMessages()).toEqual([]);
        // Server now has nothing more to send
        expect(server.getMessages(client.sessionId)).toEqual([]);
        expect(await client.getState()).toEqual(
            server.getState(client.collections),
        );
        expect(await client.getState()).toEqual({ items: expectedValue });
    });

    it('Clean client, clean server, then client does a thing', async () => {
        const client = createClient('a', ['people']);
        const server = createServer();
        const acks = server.receive(
            client.sessionId,
            await client.getMessages(),
        ); // hello world
        const clientAcks = await client.receive(
            acks.concat(server.getMessages(client.sessionId)),
        );
        // client is ack'ing the server's data, but doesn't have its own messages.
        server.receive(client.sessionId, clientAcks);
        expect(await client.getMessages()).toEqual([]);
        // Server now has nothing more to send
        expect(server.getMessages(client.sessionId)).toEqual([]);
        expect(await client.getState()).toEqual(
            server.getState(client.collections),
        );
        //
        const col = client.getCollection('people');
        await col.save('two', { name: 'yoo', age: 4 });
        await col.setAttribute('two', ['age'], 100);
        const clientMessages = await client.getMessages();
        expect(clientMessages.length).toEqual(1);
        const serverAcks = server.receive(client.sessionId, clientMessages);
        expect(serverAcks[0].type).toEqual('ack');
        expect(
            await client.receive(
                serverAcks.concat(server.getMessages(client.sessionId)),
            ),
        ).toEqual([]);
    });
});
