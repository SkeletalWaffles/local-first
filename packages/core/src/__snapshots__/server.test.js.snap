// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client-server interaction Client loses a record for some reason (??). Panic ensues. 1`] = `
Array [
  Object {
    "a": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000000:00001:aclock",
                    "map": Object {
                      "age": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                      "name": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                    },
                    "type": "map",
                  },
                  "value": Object {
                    "age": 4,
                    "name": "yoo",
                  },
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": undefined,
          "type": "sync",
        },
      ],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": -1,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltaStamp": "000000000000000:00001:aclock",
          "type": "ack",
        },
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": undefined,
          "type": "sync",
        },
      ],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": 0,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000000:00001:aclock",
                    "map": Object {
                      "age": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                      "name": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                    },
                    "type": "map",
                  },
                  "value": Object {
                    "age": 4,
                    "name": "yoo",
                  },
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 0,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
]
`;

exports[`client-server interaction Client loses a record for some reason (??). Panic ensues. 2`] = `
Array [
  Object {
    "a": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000000:00001:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000100:00000:aclock",
                    "type": "plain",
                  },
                  "value": 100,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "clientResponses": Array [],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltaStamp": "000000000000100:00000:aclock",
          "type": "ack",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": 1,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000000:00001:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000100:00000:aclock",
                    "type": "plain",
                  },
                  "value": 100,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 1,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
]
`;

exports[`client-server interaction Two clients make conflicting changes, later change wins 1`] = `
Array [
  Object {
    "a": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000000:00001:aclock",
                    "map": Object {
                      "age": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                      "name": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                    },
                    "type": "map",
                  },
                  "value": Object {
                    "age": 4,
                    "name": "yoo",
                  },
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": undefined,
          "type": "sync",
        },
      ],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": -1,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltaStamp": "000000000000000:00001:aclock",
          "type": "ack",
        },
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": undefined,
          "type": "sync",
        },
      ],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": 0,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000000:00001:aclock",
                    "map": Object {
                      "age": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                      "name": Object {
                        "hlcStamp": "000000000000000:00001:aclock",
                        "type": "plain",
                      },
                    },
                    "type": "map",
                  },
                  "value": Object {
                    "age": 4,
                    "name": "yoo",
                  },
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 0,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
]
`;

exports[`client-server interaction Two clients make conflicting changes, later change wins 2`] = `
Array [
  Object {
    "a": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000000:00001:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000100:00000:aclock",
                    "type": "plain",
                  },
                  "value": 100,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "clientResponses": Array [],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltaStamp": "000000000000100:00000:aclock",
          "type": "ack",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000000:00001:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000200:00000:bclock",
                    "type": "plain",
                  },
                  "value": 200,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 0,
          "type": "sync",
        },
      ],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": 2,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltaStamp": "000000000000200:00000:bclock",
          "type": "ack",
        },
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000000:00001:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000100:00000:aclock",
                    "type": "plain",
                  },
                  "value": 100,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 2,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": 2,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000000:00001:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000200:00000:bclock",
                    "type": "plain",
                  },
                  "value": 200,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 2,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
]
`;

exports[`client-server interaction Two clients, incremental sync 1`] = `
Array [
  Object {
    "a": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": undefined,
          "type": "sync",
        },
      ],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": -1,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": undefined,
          "type": "sync",
        },
      ],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": -1,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
]
`;

exports[`client-server interaction Two clients, incremental sync 2`] = `
Array [
  Object {
    "a": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000100:00000:aclock",
                    "map": Object {
                      "age": Object {
                        "hlcStamp": "000000000000100:00000:aclock",
                        "type": "plain",
                      },
                      "name": Object {
                        "hlcStamp": "000000000000100:00000:aclock",
                        "type": "plain",
                      },
                    },
                    "type": "map",
                  },
                  "value": Object {
                    "age": 4,
                    "name": "yoo",
                  },
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "clientResponses": Array [],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltaStamp": "000000000000100:00000:aclock",
          "type": "ack",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": 0,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000100:00000:aclock",
                    "map": Object {
                      "age": Object {
                        "hlcStamp": "000000000000100:00000:aclock",
                        "type": "plain",
                      },
                      "name": Object {
                        "hlcStamp": "000000000000100:00000:aclock",
                        "type": "plain",
                      },
                    },
                    "type": "map",
                  },
                  "value": Object {
                    "age": 4,
                    "name": "yoo",
                  },
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 0,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
]
`;

exports[`client-server interaction Two clients, incremental sync 3`] = `
Array [
  Object {
    "a": Object {
      "clientMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000100:00000:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000200:00000:aclock",
                    "type": "plain",
                  },
                  "value": 100,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": -1,
          "type": "sync",
        },
      ],
      "clientResponses": Array [],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltaStamp": "000000000000200:00000:aclock",
          "type": "ack",
        },
      ],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [
        Object {
          "collection": "people",
          "serverCursor": 1,
          "type": "ack",
        },
      ],
      "serverMessages": Array [
        Object {
          "collection": "people",
          "deltas": Array [
            Object {
              "delta": Object {
                "path": Array [
                  Object {
                    "key": "age",
                    "stamp": "000000000000100:00000:aclock",
                  },
                ],
                "type": "set",
                "value": Object {
                  "meta": Object {
                    "hlcStamp": "000000000000200:00000:aclock",
                    "type": "plain",
                  },
                  "value": 100,
                },
              },
              "node": "two",
            },
          ],
          "serverCursor": 1,
          "type": "sync",
        },
      ],
      "serverResponses": Array [],
    },
  },
  Object {
    "a": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
    "b": Object {
      "clientMessages": Array [],
      "clientResponses": Array [],
      "serverMessages": Array [],
      "serverResponses": Array [],
    },
  },
]
`;
